homeassistant/components/plugwise/sensor.py

```diff
@@ -27,374 +32,430 @@
 from .coordinator import PlugwiseDataUpdateCoordinator
 from .entity import PlugwiseEntity
 
-SENSORS: tuple[SensorEntityDescription, ...] = (
-    SensorEntityDescription(
+
+@dataclass
+class PlugwiseSensorBaseMixin:
+    """Mixin for required Plugwise sensor description keys."""
+
+    value_fn: Callable[[SmileSensors], float | int]
+
+
+@dataclass
+class PlugwiseSensorEntityDescription(SensorEntityDescription, PlugwiseSensorBaseMixin):
+    """Describes Plugwise sensor entity."""
+
+    state_class: str | None = SensorStateClass.MEASUREMENT
+
+
+SENSORS: tuple[PlugwiseSensorEntityDescription, ...] = (
+    PlugwiseSensorEntityDescription(
         key="setpoint",
         translation_key="setpoint",
         native_unit_of_measurement=UnitOfTemperature.CELSIUS,
         device_class=SensorDeviceClass.TEMPERATURE,
         state_class=SensorStateClass.MEASUREMENT,
         entity_category=EntityCategory.DIAGNOSTIC,
+        value_fn=lambda data: data["setpoint"],
```

Comment:
This doesn't really look like an improvement, we just create a lot of function objects which are not really needed.
Can you explain why it's needed?

Reply:
Hi @emontnemery - it's in preparation for strict typing, and we have a string of previous PRs going over the topic of adding/not adding it this way. For instance see https://github.com/home-assistant/core/pull/93162/files has the same approach.

If there is anything we could do differently moving to strict-typing that we missed beforehand, please let us know.

Reply:
The main problem is that the `plugwise` library is not marked as a typed package. You need to add an empty file named `py.typed` to it, as documented here https://peps.python.org/pep-0561/#packaging-type-information

Having done that, you can change plugwise/constants.py like this, after which type information will be preserved when accessing the device dictionary from the plugwise integration:

```py
BinarySensorType = Literal[
    "cooling_enabled",
    "compressor_state",
    "cooling_state",
    "dhw_state",
    "flame_state",
    "heating_state",
    "plugwise_notification",
    "slave_boiler_state",
]
BINARY_SENSORS: Final[tuple[str, ...]] = typing.get_args(BinarySensorType)
```

And the plugwise entity description like this:
```diff
@@ -23,19 +24,13 @@ from .entity import PlugwiseEntity
 SEVERITIES = ["other", "info", "warning", "error"]


-@dataclass
-class PlugwiseBinarySensorMixin:
-    """Mixin for required Plugwise binary sensor base description keys."""
-
-    value_fn: Callable[[SmileBinarySensors], bool]
-
-
 @dataclass
 class PlugwiseBinarySensorEntityDescription(
-    BinarySensorEntityDescription, PlugwiseBinarySensorMixin
+    BinarySensorEntityDescription
 ):
     """Describes a Plugwise binary sensor entity."""

+    key: BinarySensorType
     icon_off: str | None = None
```

Reply:
Sorry, in the latest version of the plugwise library, v0.31.7, the file `py.typed` is present, as are the required configuration-lines in `pyproject.toml`.

Thanks for the pointers! We didn't know that it could be solved via this method. We're still learning :)

@frenck, what about the previous PR's, should we refactor once more using the proposed method by @emontnemery?